src/dbnode/storage/index/block.go

```diff
@@ -884,6 +919,189 @@ func (b *block) addQueryResults(
+
+	if b.state == blockStateClosed {
+		return false, ErrUnableToQueryBlockClosed
+	}
+
+	size := results.Size()
+	batch := b.aggEntryPool.Get()
+	batchSize := cap(batch)
+	if batchSize == 0 {
+		batchSize = defaultAggregateResultsEntryBatchSize
+	}
+
+	aggOpts := results.AggregateResultsOptions()
+	iterateTerms := aggOpts.Type == AggregateTagNamesAndValues
+	defer func() {
+		b.aggEntryPool.Put(batch)
+	}()
+
+	iter, err := newFieldsAndTermsIterator(nil, fieldsAndTermsIteratorOpts{})
+	if err != nil {
+		return false, err
+	}
+
+	segs := b.segmentsWithRLock()
+	for _, s := range segs {
+		if opts.LimitExceeded(size) {
+			break
+		}
+
+		err = iter.Reset(s, fieldsAndTermsIteratorOpts{
+			iterateTerms: iterateTerms,
+			allowFn: func(field []byte) bool {
+				// skip any field names that we shouldn't allow.
+				if bytes.Equal(field, doc.IDReservedFieldName) {
+					return false
+				}
+				return aggOpts.TermFilter.Allow(field)
+			},
+		})
+		if err != nil {
+			return false, err
+		}
+
+		for iter.Next() {
+			if opts.LimitExceeded(size) {
+				break
+			}
+
+			field, term := iter.Current()
+			batch = b.appendFieldAndTermToBatch(batch, field, term, iterateTerms)
+			if len(batch) < batchSize {
+				continue
+			}
+
+			batch, size, err = b.addAggregateResults(cancellable, results, batch)
+			if err != nil {
+				iter.Close()
+				return false, err
+			}
+		}
+
+		if err := iter.Err(); err != nil {
+			iter.Close()
+			return false, err
+		}
+
+		if err := iter.Close(); err != nil {
+			return false, err
+		}
+	}
+
+	// NB: it is safe to Close() a fieldsAndTermsIterator twice. We structure
+	// the code to rely on this intentionally here, to ensure we're releasing
+	// any held resources by the iterator.
+	if err := iter.Close(); err != nil {
+		return false, err
+	}
+
+	// Add last batch to results if remaining.
+	if len(batch) > 0 {
+		batch, size, err = b.addAggregateResults(cancellable, results, batch)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	exhaustive := !opts.LimitExceeded(size)
+	return exhaustive, nil
+}
+
+func (b *block) appendFieldAndTermToBatch(
+	batch []AggregateResultsEntry,
+	field, term []byte,
+	includeTerms bool,
+) []AggregateResultsEntry {
+	// NB(prateek): we make a copy of the (field, term) entries returned
+	// by the iterator during traversal, because the []byte are only valid per entry during
+	// the traversal (i.e. calling Next() invalidates the []byte). We choose to do this
+	// instead of checking if the entry is required (duplicates may exist in the results map
+	// already), as it reduces contention on the map itself. Further, the ownership of these
+	// idents is transferred to the results map, which either hangs on to them (if they are new),
+	// or finalizes them if they are duplicates.
+	var (
+		entry            AggregateResultsEntry
+		lastField        []byte
+		lastFieldIsValid bool
+	)
+	// relying on the fact that iterator traversal is in order, we can avoid creating duplicate
```

Comment:
Can you explain in the comment why there might even be duplicates in the first place? I assume its because we have multiple segments which means multiple FSTs for the same field

Reply:
yeah duplicates are due to multiple FSTs. order of iteration is guaranteed as we iterate one FST at a time. 

i'll add more nuance to the comment.

Reply:
```
	// we are iterating multiple segments so we may receive duplicates (same field/term), but
	// as we are iterating one segment at a time, and because the underlying index structures
	// are FSTs, we rely on the fact that iterator traversal is in order to avoid creating duplicate
	// entries for the same fields, by checking the last batch entry to see if the bytes are
	// the same.
	// It's easier to consider an example, say we have a segment with fields/terms:
	// (f1, t1), (f1, t2), ..., (fn, t1), ..., (fn, tn)
	// as we iterate in order, we receive (f1, t1) and then (f1, t2) we can avoid the repeated f1
	// allocation if the previous entry has the same value.
	// NB: this isn't strictly true because when we switch iterating between segments,
	// the fields/terms switch in an order which doesn't have to be strictly lexicographic. In that
	// instance however, the only downside is we would be allocating more. i.e. this is just an
	// optimisation, it doesn't affect correctness.
```

Reply:
added ^^